packages/svelte-db/tests/useLiveQuery.svelte.test.ts

```diff
@@ -0,0 +1,873 @@
+          emitter.on(`sync-issue`, (changes: Array<PendingMutation>) => {
+            begin()
+            changes.forEach((change) => {
+              write({
+                key: change.key,
+                type: change.type,
+                value: change.changes as Issue,
+              })
+            })
+            commit()
+          })
+        },
+      },
+    })
+
+    // Create a transaction to perform an optimistic mutation
+    const tx = createTransaction({
+      mutationFn: async () => {
+        emitter.emit(`sync-issue`, [
+          {
+            key: `4`,
+            type: `insert`,
+            changes: {
+              id: `4`,
+              title: `New Issue`,
+              description: `New Issue Description`,
+              userId: `1`,
+            },
+          },
+        ])
+        return Promise.resolve()
+      },
+    })
+
+    cleanup = $effect.root(() => {
+      // Sync initial person data
+      emitter.emit(
+        `sync-person`,
+        initialPersons.map((person) => ({
+          key: person.id,
+          type: `insert`,
+          changes: person,
+        }))
+      )
+
+      flushSync()
+
+      // Sync initial issue data
+      emitter.emit(
+        `sync-issue`,
+        initialIssues.map((issue) => ({
+          key: issue.id,
+          type: `insert`,
+          changes: issue,
+        }))
+      )
+
+      flushSync()
+
+      // Render the hook with a query that joins persons and issues
+      const result = useLiveQuery((q) =>
+        q
+          .from({ issues: issueCollection })
+          .join({
+            type: `inner`,
+            from: { persons: personCollection },
+            on: [`@persons.id`, `=`, `@issues.userId`],
+          })
+          .select(`@issues.id`, `@issues.title`, `@persons.name`)
+          .keyBy(`@id`)
+      )
+
+      // Track each render state
+      $effect(() => {
+        renderStates.push({
+          stateSize: result.state.size,
+          hasTempKey: result.state.has(`temp-key`),
+          hasPermKey: result.state.has(`4`),
+          timestamp: Date.now(),
+        })
+      })
+
+      flushSync()
+
+      // Verify initial state
+      expect(result.state.size).toBe(3)
+
+      // Reset render states array for clarity in the remaining test
+      renderStates.length = 0
+
+      // Perform optimistic insert of a new issue
+      tx.mutate(() =>
+        issueCollection.insert(
+          {
+            id: `temp-key`,
+            title: `New Issue`,
+            description: `New Issue Description`,
+            userId: `1`,
+          },
+          { key: `temp-key` }
+        )
+      )
+
+      // Verify optimistic state is immediately reflected
+      expect(result.state.size).toBe(4)
+      expect(result.state.get(`temp-key`)).toEqual({
+        id: `temp-key`,
+        name: `John Doe`,
+        title: `New Issue`,
+      })
+    })
+
+    // Wait for the transaction to be committed
+    await tx.isPersisted.promise
```

Comment:
Since `$effect.root` can't have an async callback, we have to move this promise out of it and create different `$effect.root`s